File: src/frontend/src/api/listaComunicados.ts
import { clienteApi } from './clienteHttp'
import type { VisualizacaoComunicado } from './comunicado'
import type { components } from './tipos-gerados'

/** Um item da lista de comunicados: canal, assunto/resumo, data e visualização. */
export type ItemComunicado = {
  entregaSimuladaId: string
  mensagemId: string
  canal: string
  assuntoOuResumo: string
  criadoEm: string
  visualizacao: VisualizacaoComunicado | null
}

type CorpoProblema = Partial<components['schemas']['ProblemaListaComunicados']>

type DadosErroListaComunicados = {
  codigo: string
  correlacaoId: string | null
  ocorrencia: string
  impacto: string
  proximaAcao: string
  status: number | null
}

/** Falha tipada da lista de comunicados, com ocorrência, impacto e próxima ação segura. */
export class ErroListaComunicados extends Error {
  readonly codigo: string
  readonly correlacaoId: string | null
  readonly ocorrencia: string
  readonly impacto: string
  readonly proximaAcao: string
  readonly status: number | null

  constructor(dados: DadosErroListaComunicados) {
    super(dados.ocorrencia)
    this.name = 'ErroListaComunicados'
    this.codigo = dados.codigo
    this.correlacaoId = dados.correlacaoId
    this.ocorrencia = dados.ocorrencia
    this.impacto = dados.impacto
    this.proximaAcao = dados.proximaAcao
    this.status = dados.status
  }
}

const FALHA_INESPERADA = {
  ocorrencia: 'Os comunicados não puderam ser consultados.',
  impacto: 'A lista de comunicados pode estar desatualizada ou indisponível.',
  proximaAcao: 'Verifique se o backend local está disponível e tente novamente.',
}

const FALHA_DE_REDE: DadosErroListaComunicados = {
  codigo: 'falha_de_rede',
  correlacaoId: null,
  ocorrencia: 'Não foi possível falar com o backend local da Central Preventiva.',
  impacto: 'Nenhum comunicado pôde ser exibido.',
  proximaAcao: 'Confirme que o backend está em execução em 127.0.0.1:8000 e tente de novo.',
  status: null,
}

function erroDeResultado(erro: unknown, status: number): ErroListaComunicados {
  const corpo = (erro ?? {}) as CorpoProblema
  return new ErroListaComunicados({
    codigo: corpo.codigo ?? 'falha_inesperada',
    correlacaoId: corpo.correlacao_id ?? null,
    ocorrencia: corpo.ocorrencia ?? FALHA_INESPERADA.ocorrencia,
    impacto: corpo.impacto ?? FALHA_INESPERADA.impacto,
    proximaAcao: corpo.proxima_acao ?? FALHA_INESPERADA.proximaAcao,
    status,
  })
}

function paraItem(corpo: components['schemas']['RespostaItemComunicado']): ItemComunicado {
  return {
    entregaSimuladaId: corpo.entrega_simulada_id,
    mensagemId: corpo.mensagem_id,
    canal: corpo.canal,
    assuntoOuResumo: corpo.assunto_ou_resumo,
    criadoEm: corpo.criado_em,
    visualizacao: corpo.visualizacao
      ? { id: corpo.visualizacao.id, visualizadaEm: corpo.visualizacao.visualizada_em }
      : null,
  }
}

async function chamarLista(seguradoId: string) {
  return clienteApi.GET('/api/v1/segurados/{segurado_id}/comunicados', {
    params: { path: { segurado_id: seguradoId } },
    fetch,
  })
}

/** Lista todos os comunicados do segurado, mais antigos primeiro. */
export async function getListaComunicados(seguradoId: string): Promise<ItemComunicado[]> {
  let resultado: Awaited<ReturnType<typeof chamarLista>>
  try {
    resultado = await chamarLista(seguradoId)
  } catch {
    throw new ErroListaComunicados(FALHA_DE_REDE)
  }

  if (!resultado.response.ok || !resultado.data) {
    throw erroDeResultado(resultado.error, resultado.response.status)
  }

  return resultado.data.comunicados.map(paraItem)
}

File: src/backend/central_preventiva/adaptadores/http/lista_comunicados.py
"""Recurso REST/JSON da lista de comunicados do segurado ativo (COMUNICADOS-01, COMUNICADOS-02).

`GET /segurados/{segurado_id}/comunicados` lista todos os comunicados do segurado, mais
antigos primeiro, cada um com o estado de visualização já registrado, se houver. O detalhe e o
registro da própria visualização continuam em `visualizacao_comunicado.py` (4.3), sem
duplicação — esta rota só agrega a lista.
"""

from datetime import datetime
from uuid import UUID
from uuid import uuid4 as gerar_uuid

from fastapi import APIRouter
from fastapi.responses import JSONResponse
from pydantic import BaseModel, ConfigDict, Field

from central_preventiva.adaptadores.http.visualizacao_comunicado import (
    RespostaVisualizacaoComunicado,
)
from central_preventiva.adaptadores.persistencia.repositorio_entregas_simuladas import (
    RepositorioEntregasSimuladas,
)
from central_preventiva.adaptadores.persistencia.repositorio_visualizacoes_comunicado import (
    RepositorioVisualizacoesComunicado,
)
from central_preventiva.aplicacao.lista_comunicados import (
    ItemComunicadoListado,
    PortasListaComunicados,
    ServicoListaComunicados,
)
from central_preventiva.composicao.configuracao import Configuracao

TIPO_PROBLEMA = "application/problem+json"
CAMINHO_LISTA_COMUNICADOS = "/segurados/{segurado_id}/comunicados"


class RespostaItemComunicado(BaseModel):
    """Um item da lista de comunicados: canal, assunto/resumo, data e visualização."""

    model_config = ConfigDict(extra="forbid")

    entrega_simulada_id: UUID = Field(description="Identificador da entrega simulada.")
    mensagem_id: UUID = Field(description="Mensagem que originou o comunicado.")
    canal: str = Field(description="Canal do comunicado (`whatsapp`, `email` ou `sms`).")
    assunto_ou_resumo: str = Field(
        description="Assunto real (e-mail) ou resumo truncado do corpo (WhatsApp/SMS)."
    )
    criado_em: datetime = Field(description="Instante RFC 3339 em UTC de criação da entrega.")
    visualizacao: RespostaVisualizacaoComunicado | None = Field(
        description="Primeira visualização já registrada, ou nula se ainda não visualizado."
    )


class RespostaListaComunicados(BaseModel):
    """A lista completa de comunicados do segurado."""

    model_config = ConfigDict(extra="forbid")

    comunicados: list[RespostaItemComunicado] = Field(
        description="Todos os comunicados do segurado, mais antigos primeiro."
    )


class ProblemaListaComunicados(BaseModel):
    """Falha da lista de comunicados, com ocorrência, impacto e próxima ação segura."""

    model_config = ConfigDict(extra="forbid")

    codigo: str = Field(description="Código estável que identifica o tipo da falha.")
    correlacao_id: str = Field(description="Identificador único desta ocorrência de falha.")
    ocorrencia: str = Field(description="O que aconteceu, em português brasileiro.")
    impacto: str = Field(description="Efeito prático da falha para quem consultou o recurso.")
    proxima_acao: str = Field(description="Próxima ação segura recomendada para contornar a falha.")


def _problema_identificador_invalido() -> JSONResponse:
    """Resposta de identificador que não é um UUID válido."""

    corpo = ProblemaListaComunicados(
        codigo="identificador_invalido",
        correlacao_id=str(gerar_uuid()),
        ocorrencia="O identificador do segurado não é um UUID válido.",
        impacto="Nenhum comunicado pode ser exibido.",
        proxima_acao="Consulte a lista pelo identificador do segurado retornado pela API.",
    )
    return JSONResponse(status_code=422, media_type=TIPO_PROBLEMA, content=corpo.model_dump())


def _resposta_visualizacao(
    visualizacao: ItemComunicadoListado,
) -> RespostaVisualizacaoComunicado | None:
    """Traduz a visualização do item para o contrato público, ou devolve nula."""

    if visualizacao.visualizacao is None:
        return None
    return RespostaVisualizacaoComunicado(
        id=visualizacao.visualizacao.id, visualizada_em=visualizacao.visualizacao.visualizada_em
    )


def _resposta_item(item: ItemComunicadoListado) -> RespostaItemComunicado:
    """Traduz um item da lista para o contrato público."""

    return RespostaItemComunicado(
        entrega_simulada_id=item.entrega_simulada_id,
        mensagem_id=item.mensagem_id,
        canal=str(item.canal),
        assunto_ou_resumo=item.assunto_ou_resumo,
        criado_em=item.criado_em,
        visualizacao=_resposta_visualizacao(item),
    )


def montar_servico_lista_comunicados(configuracao: Configuracao) -> ServicoListaComunicados:
    """Compõe o serviço de lista sobre os repositórios reais do DuckDB local."""

    caminho = configuracao.caminho_banco
    return ServicoListaComunicados(
        PortasListaComunicados(
            entregas=RepositorioEntregasSimuladas(caminho),
            visualizacoes=RepositorioVisualizacoesComunicado(caminho),
        )
    )


def criar_roteador(configuracao: Configuracao) -> APIRouter:
    """Compõe o recurso da lista de comunicados do segurado ativo."""

    roteador = APIRouter(tags=["Comunicado"])
    servico = montar_servico_lista_comunicados(configuracao)

    @roteador.get(
        CAMINHO_LISTA_COMUNICADOS,
        response_model=RespostaListaComunicados,
        status_code=200,
        summary="Listar todos os comunicados do segurado ativo",
        description=(
            "Devolve todos os comunicados do segurado, mais antigos primeiro, com canal, "
            "assunto/resumo, data e a visualização já registrada, se houver. Lista vazia "
            "quando o segurado não tiver nenhum comunicado."
        ),
        responses={
            200: {"description": "Consulta realizada, com ou sem comunicados."},
            422: {
                "description": "O identificador do segurado não é um UUID válido.",
                "model": ProblemaListaComunicados,
            },
        },
    )
    async def listar_comunicados(segurado_id: str) -> RespostaListaComunicados | JSONResponse:  # pyright: ignore[reportUnusedFunction]
        """Traduz `ServicoListaComunicados.listar` para o contrato público."""

        try:
            segurado_uuid = UUID(segurado_id)
        except ValueError:
            return _problema_identificador_invalido()

        itens = servico.listar(segurado_uuid)
        return RespostaListaComunicados(comunicados=[_resposta_item(item) for item in itens])

    return roteador
